packages/core/src/routes/experience/classes/verifications/one-time-token-verification.ts

```diff
@@ -0,0 +1,112 @@
+import { type ToZodObject } from '@logto/connector-kit';
+import {
+  type InteractionIdentifier,
+  oneTimeTokenVerificationVerifyPayloadGuard,
+  type SignInIdentifier,
+  type User,
+  VerificationType,
+} from '@logto/schemas';
+import { generateStandardId } from '@logto/shared';
+import { z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import type Libraries from '#src/tenants/Libraries.js';
+import type Queries from '#src/tenants/Queries.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import { findUserByIdentifier } from '../utils.js';
+
+import { type IdentifierVerificationRecord } from './verification-record.js';
+
+export type OneTimeTokenVerificationRecordData = {
+  id: string;
+  type: VerificationType.OneTimeToken;
+  identifier: InteractionIdentifier<SignInIdentifier.Email>;
+  verified: boolean;
+};
+
+export const oneTimeTokenVerificationRecordDataGuard =
+  oneTimeTokenVerificationVerifyPayloadGuard.extend({
+    id: z.string(),
+    type: z.literal(VerificationType.OneTimeToken),
+    verified: z.boolean(),
+  }) satisfies ToZodObject<OneTimeTokenVerificationRecordData>;
+
+export class OneTimeTokenVerification
+  implements IdentifierVerificationRecord<VerificationType.OneTimeToken>
+{
+  /** Factory method to create a new `OneTimeTokenVerification` record using an identifier */
+  static create(
+    libraries: Libraries,
+    queries: Queries,
+    identifier: InteractionIdentifier<SignInIdentifier.Email>
+  ) {
+    return new OneTimeTokenVerification(libraries, queries, {
+      id: generateStandardId(),
+      type: VerificationType.OneTimeToken,
+      identifier,
+      verified: false,
+    });
+  }
+
+  readonly type = VerificationType.OneTimeToken;
+  readonly id: string;
+  private readonly identifier: InteractionIdentifier<SignInIdentifier.Email>;
+  private verified: boolean;
+
+  /**
+   * The constructor method is intended to be used internally by the interaction class
+   * to instantiate a `VerificationRecord` object from existing `OneTimeTokenVerificationRecordData`.
+   * It directly sets the instance properties based on the provided data.
+   * For creating a new verification record from context, use the static `create` method instead.
+   */
+  constructor(
+    private readonly libraries: Libraries,
+    private readonly queries: Queries,
+    data: OneTimeTokenVerificationRecordData
+  ) {
+    const { id, identifier, verified } = data;
+
+    this.id = id;
+    this.identifier = identifier;
+    this.verified = verified;
+  }
+
+  /** Returns true if a userId is set */
+  get isVerified() {
+    return this.verified;
+  }
+
+  /**
+   * Verifies if the one-time token matches the record in database with the provided email.
+   */
+  async verify(token: string) {
+    await this.libraries.oneTimeTokens.verifyOneTimeToken(token, this.identifier.value);
+    this.verified = true;
+  }
```

Comment:
not sure, but would it be more handy to hide this `verify` method? Instead, always require `token` in the create method, and verify the token on instance creation.  Once created the verify status won't change.  Also only verified token can successfully initiate a one-time-token verification.

Reply:
OK, that works too. One-time tokens are not as sensitive as passwords, so ideally we can store them in session.

But doesn't this introduce a new verification pattern? I thought having the `verify()` function was a preferred coding pattern. I personally prefer having `create` and `verify` functions independent from each other, and this works better with the `withSentinel` middleware since I can only call `verify` in the middleware.

Reply:
Make sense, considering we might have SIE-triggered one-time-token verification shortly, better to have them predefined.